### PR TITLE
Support for selection and filter on no-dictionary columns

### PR DESCRIPTION
**Pull request by kishoreg**
Preliminary changes to support filters and selection on no dictionary columns. As of now, only EqualsPredicateEvaluator is supported. Submitting this to get early feedback. Will add support for all datatypes and predicate types in the next update.

### CODE REVIEW

**Comment by mayankshriv** [comment]
Added another commit to the PR to complete the feature across all predicates for all data types.
    
1. Enhanced predicates (NotEquals, In, NotIn and Range) to support
no-dictionary based predicate evaluation.
    
2. Added support selection queries with no-dictionary indices.
    
3. Added unit test for all no-dictionary based predicate evaluators for
all supported data types.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/predicate/EqualsPredicateEvaluatorFactory.java [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+
+import com.linkedin.pinot.common.data.FieldSpec;
+import com.linkedin.pinot.core.common.predicate.EqPredicate;
+import com.linkedin.pinot.core.segment.index.readers.Dictionary;
+
+
+/**
+ * Factory for equality predicate evaluators.
+ */
+public class EqualsPredicateEvaluatorFactory {
```
Comment (kishoreg) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/predicate/EqualsPredicateEvaluatorFactory.java]: Thanks for adding this.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/predicate/RangePredicateEvaluatorFactory.java [hunk 2]
```diff
@@ -0,0 +1,557 @@
...
+      case STRING:
+        return new StringNoDictionaryBasedRangeEvaluator(predicate);
+
+      default:
+        throw new UnsupportedOperationException(
+            "No dictionary based Equals predicate evaluator not supported for datatype:" + dataType);
+    }
+  }
+
+  public static class OfflineDictionaryBasedPredicateEvaluator extends BasePredicateEvaluator {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/predicate/RangePredicateEvaluatorFactory.java]: private?

File: pinot-core/src/main/java/com/linkedin/pinot/core/plan/FilterPlanNode.java [hunk 3]
```diff
@@ -126,8 +126,9 @@ public static BaseFilterOperator constructPhysicalOperator(FilterQueryTree filte
       BaseFilterOperator baseFilterOperator;
       int startDocId = 0;
       int endDocId = segment.getSegmentMetadata().getTotalRawDocs() - 1; //end is inclusive
-      if (dataSourceMetadata.hasInvertedIndex()) {
-        // range/REGEX evaluation based on inv index is inefficient, so do this only if is NOT range/regex.
+      //use inverted index only if the column has dictionary.
+      if (dataSourceMetadata.hasInvertedIndex() && dataSourceMetadata.hasDictionary()) {
+        // range evaluation based on inv index is inefficient, so do this only if is NOT range.
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/plan/FilterPlanNode.java]: Update comment on REGEXP_LIKE

Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/plan/FilterPlanNode.java]: Fixed.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/predicate/RangePredicateEvaluatorFactory.java (hunk 2 above)
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/predicate/RangePredicateEvaluatorFactory.java]: Fixed.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/predicate/PredicateEvaluatorProvider.java [hunk 4]
```diff
@@ -26,31 +28,57 @@
...
+        return NotInPredicateEvaluatorFactory.newNoDictionaryBasedEvaluator((NotInPredicate) predicate, dataType);
+      case RANGE:
+        if (dictionary instanceof ImmutableDictionaryReader) {
+          return RangePredicateEvaluatorFactory.newNoDictionaryBasedEvaluator(
+              (RangePredicate) predicate, dataType);
+        } else {
+          return RangePredicateEvaluatorFactory.newNoDictionaryBasedEvaluator((RangePredicate) predicate, dataType);
+        }
+      case REGEXP_LIKE:
+        return  RegexpLikePredicateEvaluatorFactory.newNoDictionaryBasedEvaluator((RegexpLikePredicate) predicate, dictionary);
```
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/predicate/PredicateEvaluatorProvider.java]: Removed.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/predicate/PredicateEvaluatorProvider.java [hunk 5]
```diff
@@ -26,31 +28,57 @@
...
-
 public class PredicateEvaluatorProvider {
 
-  public static PredicateEvaluator getPredicateFunctionFor(Predicate predicate, Dictionary dictionary) {
-    switch (predicate.getType()) {
+  public static PredicateEvaluator getPredicateFunctionFor(final Predicate predicate, DataSource dataSource) {
+    Dictionary dictionary = dataSource.getDictionary();
+    if (dataSource.getDataSourceMetadata().hasDictionary()) {
+      switch (predicate.getType()) {
       case EQ:
```
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/predicate/PredicateEvaluatorProvider.java]: Fixed.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/predicate/PredicateEvaluatorProvider.java [hunk 6]
```diff
@@ -26,31 +28,57 @@
 import com.linkedin.pinot.core.segment.index.readers.Dictionary;
 import com.linkedin.pinot.core.segment.index.readers.ImmutableDictionaryReader;
 
-
 public class PredicateEvaluatorProvider {
 
-  public static PredicateEvaluator getPredicateFunctionFor(Predicate predicate, Dictionary dictionary) {
-    switch (predicate.getType()) {
+  public static PredicateEvaluator getPredicateFunctionFor(final Predicate predicate, DataSource dataSource) {
+    Dictionary dictionary = dataSource.getDictionary();
```
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/predicate/PredicateEvaluatorProvider.java]: Done.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/predicate/NotInPredicateEvaluatorFactory.java [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+      default:
+        throw new UnsupportedOperationException(
+            "No dictionary based Equals predicate evaluator not supported for datatype:" + dataType);
+    }
+  }
+
+  /**
+   * Dictionary based implementation of not-in predicate evaluator.
+   */
+  public static class DictionaryBasedNotInPredicateEvaluator extends BasePredicateEvaluator {
```
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/predicate/NotInPredicateEvaluatorFactory.java]: Fixed.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/predicate/NotEqualsPredicateEvaluatorFactory.java [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+      default:
+        throw new UnsupportedOperationException("Data type not supported for index without dictionary: " + dataType);
+    }
+  }
+
+  /**
+   *
+   * No dictionary NEQ evaluator for int data type.
+   */
+  public static final class IntNoDictionaryNeqEvaluator extends BasePredicateEvaluator {
```
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/predicate/NotEqualsPredicateEvaluatorFactory.java]: Fixed.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/predicate/EqualsPredicateEvaluatorFactory.java [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+        throw new UnsupportedOperationException(
+            "No dictionary based Equals predicate evaluator not supported for datatype:" + dataType);
+    }
+  }
+
+  /**
+   *
+   * No dictionary equality evaluator for int data type.
+   */
+  public static final class IntNoDictionaryBasedEqualsEvaluator extends BasePredicateEvaluator {
```
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/predicate/EqualsPredicateEvaluatorFactory.java]: Done.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/dociditerators/SVScanDocIdIterator.java [hunk 10]
```diff
@@ -152,4 +170,91 @@ public MutableRoaringBitmap applyAnd(MutableRoaringBitmap answer) {
...
+      default:
+        throw new UnsupportedOperationException("Index without dictionary not supported for data type: " + dataType);
+
+    }
+    return valueMatcher;
+  }
+}
+
+abstract class ValueMatcher {
+  protected PredicateEvaluator evaluator;
```
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/dociditerators/SVScanDocIdIterator.java]: Done.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/dociditerators/SVScanDocIdIterator.java [hunk 11]
```diff
@@ -152,4 +170,91 @@ public MutableRoaringBitmap applyAnd(MutableRoaringBitmap answer) {
...
+
+      default:
+        throw new UnsupportedOperationException("Index without dictionary not supported for data type: " + dataType);
+
+    }
+    return valueMatcher;
+  }
+}
+
+abstract class ValueMatcher {
```
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/dociditerators/SVScanDocIdIterator.java]: Done.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/dociditerators/SVScanDocIdIterator.java [hunk 12]
```diff
@@ -140,7 +155,10 @@ public MutableRoaringBitmap applyAnd(MutableRoaringBitmap answer) {
       if (docId >= startDocId) {
         valueIterator.skipTo(docId);
         _numEntriesScanned++;
-        if (evaluator.apply(valueIterator.nextIntVal())) {
+        // if (evaluator.apply(valueIterator.nextIntVal())) {
+        // result.add(docId);
+        // }
```
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/dociditerators/SVScanDocIdIterator.java]: Done.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/dociditerators/SVScanDocIdIterator.java [hunk 13]
```diff
@@ -101,17 +113,20 @@ public int advance(int targetDocId) {
...
-        // dictIdForCurrentDoc);
+      if (valueMatcher.doesCurrentEntryMatch(valueIterator)) {
         return currentDocId;
       }
+      // int dictIdForCurrentDoc = valueIterator.nextIntVal();
+      // if (evaluator.apply(dictIdForCurrentDoc)) {
+      // // System.out.println("Returning deom " + this +"doc Id:"+ currentDocId + " dictId:"+
+      // // dictIdForCurrentDoc);
+      // return currentDocId;
+      // }
```
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/dociditerators/SVScanDocIdIterator.java]: Done.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/dociditerators/SVScanDocIdIterator.java [hunk 14]
```diff
@@ -75,8 +86,9 @@ public boolean isMatch(int docId) {
     }
     valueIterator.skipTo(docId);
     _numEntriesScanned++;
-    int dictIdForCurrentDoc = valueIterator.nextIntVal();
-    return evaluator.apply(dictIdForCurrentDoc);
+    // int dictIdForCurrentDoc = valueIterator.nextIntVal();
+    // return evaluator.apply(dictIdForCurrentDoc);
```
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/dociditerators/SVScanDocIdIterator.java]: Done.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/dociditerators/SVScanDocIdIterator.java [hunk 15]
```diff
@@ -47,10 +49,18 @@ public SVScanDocIdIterator(String datasourceName, BlockValSet blockValSet, Block
       setStartDocId(blockMetadata.getStartDocId());
       setEndDocId(blockMetadata.getEndDocId());
     }
+
+    if (blockMetadata.hasDictionary()) {
+      valueMatcher = new IntMatcher();
```
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/dociditerators/SVScanDocIdIterator.java]: Done.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/dociditerators/SVScanDocIdIterator.java [hunk 16]
```diff
@@ -24,21 +26,21 @@
...
 public class SVScanDocIdIterator implements ScanBasedDocIdIterator {
   int currentDocId = -1;
   BlockSingleValIterator valueIterator;
   private int startDocId;
   private int endDocId;
-  private PredicateEvaluator evaluator;
+  protected PredicateEvaluator evaluator;
   private String datasourceName;
   private int _numEntriesScanned = 0;
+  ValueMatcher valueMatcher;
```
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/dociditerators/SVScanDocIdIterator.java]: Done.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/dociditerators/SVScanDocIdIterator.java [hunk 17]
```diff
@@ -24,21 +26,21 @@
...
 import com.linkedin.pinot.core.operator.filter.predicate.PredicateEvaluator;
 
-
 public class SVScanDocIdIterator implements ScanBasedDocIdIterator {
   int currentDocId = -1;
   BlockSingleValIterator valueIterator;
   private int startDocId;
   private int endDocId;
-  private PredicateEvaluator evaluator;
+  protected PredicateEvaluator evaluator;
```
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/dociditerators/SVScanDocIdIterator.java]: Made private.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/dociditerators/SVScanDocIdIterator.java [hunk 18]
```diff
@@ -24,21 +26,21 @@
 import com.linkedin.pinot.core.common.Constants;
 import com.linkedin.pinot.core.operator.filter.predicate.PredicateEvaluator;
 
-
 public class SVScanDocIdIterator implements ScanBasedDocIdIterator {
   int currentDocId = -1;
   BlockSingleValIterator valueIterator;
```
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/dociditerators/SVScanDocIdIterator.java]: Done.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/dociditerators/SVScanDocIdIterator.java [hunk 19]
```diff
@@ -24,21 +26,21 @@
 import com.linkedin.pinot.core.common.Constants;
 import com.linkedin.pinot.core.operator.filter.predicate.PredicateEvaluator;
 
-
 public class SVScanDocIdIterator implements ScanBasedDocIdIterator {
   int currentDocId = -1;
```
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/dociditerators/SVScanDocIdIterator.java]: Done.

**Comment by mayankshriv** [comment]
@Jackie-Jiang Addressed all the comments. You are correct, no-dictionary index is currently not supported for multi-valued columns.

**Pull request merged**